prov/efa/src/efa_av.c

```diff
@@ -565,68 +593,75 @@ static int efa_av_remove(struct fid_av *av_fid, fi_addr_t *fi_addr,
 	struct rxr_ep *rxr_ep;
 
 	av = container_of(av_fid, struct efa_av, util_av.av_fid);
-	if (av->ep_type == FI_EP_DGRAM) {
-		for (i = 0; i < count; i++) {
-			ret = efa_av_remove_ah(&av->util_av.av_fid, &fi_addr[i],
-					       1, flags);
-			if (ret)
-				goto out;
-		}
-		goto out;
-	}
+	if (av->ep_type == FI_EP_DGRAM)
+		return efa_dgram_av_remove(av_fid, fi_addr, count, flags);
+
 	fastlock_acquire(&av->util_av.lock);
 	/* currently multiple EP bind to same av is not supported */
 	rxr_ep = container_of(av->util_av.ep_list.next, struct rxr_ep, util_ep.av_entry);
 	for (i = 0; i < count; i++) {
-		if (fi_addr[i] == FI_ADDR_NOTAVAIL ||
-		    fi_addr[i] > av->count) {
+		if (fi_addr[i] == FI_ADDR_NOTAVAIL || fi_addr[i] > av->count) {
 			ret = -FI_ENOENT;
-			goto release_lock;
+			goto out;
 		}
-		util_av_entry = ofi_bufpool_get_ibuf(
-						av->util_av.av_entry_pool,
-						fi_addr[i]);
+
+		util_av_entry = ofi_bufpool_get_ibuf(av->util_av.av_entry_pool, fi_addr[i]);
 		if (!util_av_entry) {
 			ret = -FI_ENOENT;
-			goto release_lock;
+			goto out;
 		}
-		av_entry = (struct efa_av_entry *)util_av_entry->data;
+
 		peer = rxr_ep_get_peer(rxr_ep, fi_addr[i]);
 
-		/* Check if the peer is in use if it is then return */
 		ret = efa_peer_in_use(peer);
 		if (ret)
-			goto release_lock;
+			goto out;
 
-		/* Only if the peer is not in use reset the peer */
 		if (peer->rx_init)
 			efa_peer_reset(peer);
 
-		ret = efa_av_remove_ah(&av->util_av.av_fid, &fi_addr[i], 1,
+		/*
+		 * Clearing the 3 resources of an av entry:
+		 *
+		 *     address handler (AH), shm_av_entry and util_av_entry.
+		 *
+		 * We will try our best to remove these resources. If releasing one
+		 * resource failed, we will not stop. Instead we save the error code
+		 * and continue to release other resources.
+		 */
+		av_entry = (struct efa_av_entry *)util_av_entry->data;
+		err = efa_av_remove_ah(&av->util_av.av_fid, &fi_addr[i], 1,
 				       flags);
+		if (err)
+			ret = err;
+
+		if (av_entry->local_mapping) {
+			err = fi_av_remove(av->shm_rdm_av, &av_entry->shm_rdm_addr, 1, flags);
+			if (err) {
+				ret = err;
+			} else {
+				av->shm_used--;
+				assert(av_entry->shm_rdm_addr < rxr_env.shm_av_size);
+				av->shm_rdm_addr_map[av_entry->shm_rdm_addr] = FI_ADDR_UNSPEC;
+			}
+		}
+
+		ofi_av_remove_addr(&av->util_av, *fi_addr);
```

Comment:
do we need to check the return status of `ofi_av_remove_addr`?

Reply:
+1

Reply:
added in new revision